src/vos/vos_dtx.c

```diff
@@ -717,6 +730,469 @@ vos_dtx_abort_one(struct vos_container *cont, struct daos_tx_id *dti,
+		return false;
+
+	return true;
+}
+
+/* The caller has started PMDK transaction. */
+int
+vos_dtx_register_record(struct umem_instance *umm, umem_id_t record,
+			uint32_t type, uint32_t flags)
+{
+	struct daos_tx_handle		*dth = vos_dth_get();
+	struct vos_dtx_entry_df		*dtx;
+	struct vos_dtx_record_df	*rec;
+	struct vos_dtx_share		*vds;
+	struct vos_dtx_share		*next;
+	umem_id_t			 rec_mmid = UMMID_NULL;
+	umem_id_t			*entry = NULL;
+	uint32_t			*shares = NULL;
+	int				 rc = 0;
+	bool				 shared = false;
+
+	switch (type) {
+	case DTX_RT_OBJ: {
+		struct vos_obj_df	*obj;
+
+		obj = umem_id2ptr(umm, record);
+		entry = &obj->vo_dtx;
+		if (dth == NULL || dth->dth_intent == DAOS_INTENT_UPDATE)
+			shares = &obj->vo_dtx_shares;
+		umem_tx_add_ptr(umm, obj, sizeof(*obj));
+		break;
+	}
+	case DTX_RT_KEY: {
+		struct vos_krec_df	*key;
+
+		key = umem_id2ptr(umm, record);
+		entry = &key->kr_dtx;
+		if (dth == NULL || dth->dth_intent == DAOS_INTENT_UPDATE)
+			shares = &key->kr_dtx_shares;
+		umem_tx_add_ptr(umm, key, sizeof(*key));
+		break;
+	}
+	case DTX_RT_SVT: {
+		struct vos_irec_df	*svt;
+
+		svt = umem_id2ptr(umm, record);
+		entry = &svt->ir_dtx;
+		umem_tx_add_ptr(umm, entry, sizeof(*entry));
+		break;
+	}
+	case DTX_RT_EVT: {
+		struct evt_desc		*evt;
+
+		evt = umem_id2ptr(umm, record);
+		entry = &evt->dc_dtx;
+		umem_tx_add_ptr(umm, entry, sizeof(*entry));
+		break;
+	}
+	default:
+		D_ERROR("Unknown DTX type %u\n", type);
+		return -DER_INVAL;
+	}
+
+	if (dth == NULL) {
+		*entry = UMMID_NULL;
+		if (shares != NULL)
+			*shares = 0;
+
+		return 0;
+	}
+
+	rec_mmid = umem_zalloc(umm, sizeof(struct vos_dtx_record_df));
+	if (UMMID_IS_NULL(rec_mmid))
+		return -DER_NOMEM;
+
+	rec = umem_id2ptr(umm, rec_mmid);
+	umem_tx_add_ptr(umm, rec, sizeof(*rec));
+
+	rec->tr_type = type;
+	rec->tr_flags = flags;
+	rec->tr_record = record;
+	rec->tr_next = UMMID_NULL;
+
+	if (UMMID_IS_NULL(dth->dth_ent)) {
+		struct vos_container	*cont;
+		umem_id_t		 dtx_mmid;
+		daos_iov_t		 kiov;
+		daos_iov_t		 riov;
+		struct dtx_rec_bundle	 rbund;
+
+		D_ASSERT(flags == 0);
+
+		cont = vos_hdl2cont(dth->dth_coh);
+		D_ASSERT(cont != NULL);
+
+		dtx_mmid = umem_zalloc(umm, sizeof(struct vos_dtx_entry_df));
+		if (UMMID_IS_NULL(dtx_mmid))
+			return -DER_NOMEM;
+
+		dtx = umem_id2ptr(umm, dtx_mmid);
+		umem_tx_add_ptr(umm, dtx, sizeof(*dtx));
+
+		dtx->te_xid = dth->dth_xid;
+		dtx->te_oid = dth->dth_oid;
+		if (dth->dth_dkey != NULL) {
+			dtx->te_dkey_hash[0] = d_hash_murmur64(
+						dth->dth_dkey->iov_buf,
+						dth->dth_dkey->iov_len,
+						VOS_BTR_MUR_SEED);
+			dtx->te_dkey_hash[1] = d_hash_string_u32(
+						dth->dth_dkey->iov_buf,
+						dth->dth_dkey->iov_len);
+		} else {
+			dtx->te_dkey_hash[0] = 0;
+			dtx->te_dkey_hash[1] = 0;
+		}
+		dtx->te_epoch = dth->dth_epoch;
+		dtx->te_ver = dth->dth_ver;
+		dtx->te_state = DTX_ST_INIT;
+		dtx->te_flags = 0;
+		dtx->te_intent = dth->dth_intent;
+		dtx->te_sec = time(NULL);
```

Comment:
Is this for getting a wall clock or a monotonic clock? If we want a monotonic clock here, I think we'd call clock_gettime(CLOCK_MONOTONIC/CLOCK_MONOTONIC_COARSE, &dtx->te_sec), it can ensure no syscall and monotonicity of the returned time.

Please check all the time(NULL) used in many other places.

Reply:
We need a wall clock, that is helpful for debug.